### ISSUE TITLE
LoRA fine-tune on LIBERO-Spatial: 0% success, mode collapse — missing clip_grad_norm_ in finetune.py?

### ISSUE THREAD

**Issue by shant0602** [openvla/openvla#333]
**Hardware:** Lambda Labs GH200 (ARM64, H100, CUDA 12.3, PyTorch 2.2)
**Config:** LoRA rank=32, LR=5e-4, batch=16, 10k steps on `libero_spatial_no_noops`
**Base model:** `openvla/openvla-7b`

## Results

| Model | Eval (LIBERO-Spatial, 10 tasks × 5 trials) | Notes |
|---|---|---|
| `openvla/openvla-7b-finetuned-libero-spatial` (authors' checkpoint) | **37/50 = 74%** | Paper reports 84.7% on x86; gap likely due to flash-attn not installed (pre-built wheel in NVIDIA container is x86-only; we did not compile from source) |
| Our LoRA fine-tune (base model + `finetune.py`, 10k steps) | **0/50 = 0%** | Mode collapse — identical action tokens every step regardless of observation |

Both models evaluated on the **exact same hardware and setup**, confirming the eval environment is not the issue.

## Training curves (W&B)

![Training curves](https://raw.githubusercontent.com/shant0602/PhysicalAI/feature/openvla-training-pipeline/assets/training_curves_run1.png)

**Key observations:**
- Loss spike at step ~6000: jumps from ~2.2 → ~5.2
- Action accuracy crashes to near 0 at the same step and never recovers
- Final action accuracy: 0.39 (expected ~0.90+)

## Symptoms

- `action_accuracy` reaches only ~0.39 at 10k steps (expected ~0.9+)
- `train_loss` never meaningfully converges
- At inference, model outputs **identical action tokens every step regardless of image input** — mode collapse

## Hypothesis

`finetune.py` uses raw AdamW with no gradient clipping. A single outlier batch can produce an unclamped gradient that corrupts the LoRA weights. Standard transformer fine-tuning (LLaMA, PrismaticVLM, etc.) uses `torch.nn.utils.clip_grad_norm_(params, max_norm=1.0)` before the optimizer step — this is absent from `finetune.py`.

Note: multiple other users have reported the same 0% success rate symptom in #299 (open since Sep 2025, no fix yet).

## Question

Was gradient clipping used in the internal training pipeline for the published LIBERO checkpoints? If so, would you accept a PR adding it to `finetune.py`?

## Proposed fix (one line)

```python
# Around line 316 of vla-scripts/finetune.py, before optimizer.step():
torch.nn.utils.clip_grad_norm_(trainable_params, max_norm=1.0)
optimizer.step()
```

**Comment by shant0602** [comment]
cc @moojink @kpertsch — posting results in case this is useful for the repo.

**Update: fix confirmed with experimental results — PR #334**

We ran two training runs on the same hardware (GH200, ARM64, CUDA 12.3) to isolate the effect of gradient clipping.

### Training runs comparison

| | Run 1 (original) | Run 2 (with fix) |
|---|---|---|
| Gradient clipping | No | Yes |
| Start checkpoint | openvla/openvla-7b | openvla/openvla-7b |
| Steps | 10k | 20k (two sequential 10k runs, second resumed from run 1's merged checkpoint) |
| Batch size | 16 | 16 |
| LoRA rank | 32 | 32 |
| Learning rate | 5e-4 | 5e-4 |
| Final train loss | 2.66 | 0.68 |
| Final action acc (training) | 0.39 | 0.82 |
| Loss spike? | Yes (~step 6k, 2.2 → 5.2) | No |

### LIBERO-Spatial evaluation (10 tasks × 5 trials = 50 episodes)

| Checkpoint | Steps | Score (50 eps) | Notes |
|---|---|---|---|
| `openvla/openvla-7b-finetuned-libero-spatial` (authors') | 10k | 37/50 = 74% | Baseline on same ARM64 hardware |
| Run 1 — no clipping | 10k | 0/50 = 0% | Mode collapse — identical action tokens at inference regardless of observation |
| **Run 2 — with clipping** | **20k** | **36/50 = 72%** | Merged checkpoint, public on HF (note: 2× the training budget of the official checkpoint) |

### Training curves — before and after

| ✗ Run 1 — no gradient clipping (10k steps) | ✓ Run 2 — with gradient clipping (20k steps) |
|---|---|
| ![Run 1 loss spike](https://raw.githubusercontent.com/shant0602/PhysicalAI/feature/openvla-training-pipeline/assets/training_curves_run1.png) | ![Run 2 clean convergence](https://raw.githubusercontent.com/shant0602/PhysicalAI/feature/openvla-training-pipeline/assets/training_curves_20k.png) |
| Loss spikes at step ~6k, accuracy collapses to 0 and never recovers | Loss decreases steadily 2.4 → 0.68, accuracy rises 0.05 → 0.82 — no spike |

### Notes on the gap vs. paper (84.7%)

| Factor | Our setup | Paper setup |
|---|---|---|
| Hardware | GH200, ARM64 | x86 |
| flash-attn | Built from source (2.5.8) | Pre-built wheel |
| Official checkpoint eval on our hardware | 74% | 84.7% (paper) |

The ~12pp gap is consistent with the hardware difference — the official checkpoint also scores 74% (not 84.7%) on our ARM64 setup, suggesting hardware is the primary driver of the gap rather than training quality.

### Artifacts

| | Link |
|---|---|
| PR with one-line fix | https://github.com/openvla/openvla/pull/334 |
| 20k checkpoint (public, merged weights) | https://huggingface.co/shant0602/openvla-7b-libero-spatial-lora-20k |

**Comment by rylinjames** [comment]
your hypothesis is almost certainly right. the loss spike at step ~6000 jumping 2.2→5.2 with action accuracy crashing to ~0 is textbook unclipped gradient blowup on lora, especially at lr=5e-4 rank=32 which is on the aggressive side. adding `torch.nn.utils.clip_grad_norm_(params, max_norm=1.0)` before `optimizer.step()` in finetune.py should fix it, and dropping lr to 2e-4 as a belt-and-suspenders. also worth enabling bf16 instead of fp16 if you're on gh200 (hopper), fp16 on libero has a history of flaky lora. reflex doesn't cover openvla training (we focus on deploy for smolvla/pi0/pi0.5/gr00t), so this isn't our layer, but if you end up with a working lora checkpoint and want onnx + trt deploy, we export with cos=+1.0 parity. github.com/rylinjames/reflex-vla has good-first-issues, or dm romirj on discord / playindus@gmail.com for collaborator access.
